lib/Dialect/ESI/cosim/cosim_dpi_server/Server.cpp

```diff
@@ -0,0 +1,162 @@
+/// others.
+kj::Promise<void> EndpointServer::send(SendContext context) {
+  KJ_REQUIRE(open, "EndPoint closed already");
+  auto capnpMsgPointer = context.getParams().getMsg();
+  KJ_REQUIRE(capnpMsgPointer.isStruct(),
+             "Only messages can go in the 'msg' parameter");
+
+  // Copy the incoming message into a flat, single segment buffer.
+  auto msgSize = capnpMsgPointer.targetSize();
+  auto builder = std::make_unique<MallocMessageBuilder>(
+      msgSize.wordCount + 1, AllocationStrategy::FIXED_SIZE);
+  builder->setRoot(capnpMsgPointer);
+  auto segments = builder->getSegmentsForOutput();
+  KJ_ASSERT(segments.size() == 1);
+
+  // Now copy it into a blob and queue it.
+  auto fstSegmentData = segments[0].asBytes();
+  auto blob = std::make_shared<Endpoint::Blob>(fstSegmentData.begin(),
+                                               fstSegmentData.end());
+  endpoint.pushMessageToSim(blob);
+  return kj::READY_NOW;
+}
+
+kj::Promise<void> EndpointServer::close(CloseContext context) {
+  KJ_REQUIRE(open, "EndPoint closed already");
+  open = false;
+  endpoint.returnForUse();
+  return kj::READY_NOW;
+}
+
+/// ----- CosimServer definitions.
+
+CosimServer::CosimServer(EndpointRegistry *reg) : reg(reg) {}
+
+kj::Promise<void> CosimServer::list(ListContext context) {
+  auto ifaces = context.getResults().initIfaces((unsigned int)reg->size());
+  unsigned int ctr = 0u;
+  reg->iterateEndpoints([&](int id, const Endpoint &ep) {
+    ifaces[ctr].setEndpointID(id);
+    ifaces[ctr].setSendTypeID(ep.getSendTypeId());
+    ifaces[ctr].setRecvTypeID(ep.getRecvTypeId());
+    ctr++;
+  });
+  return kj::READY_NOW;
+}
+
+kj::Promise<void> CosimServer::open(OpenContext ctxt) {
+  Endpoint *ep;
+  bool found = reg->get(ctxt.getParams().getIface().getEndpointID(), ep);
+  KJ_REQUIRE(found, "Could not find endpoint");
+
+  auto gotLock = ep->setInUse();
+  KJ_REQUIRE(gotLock, "Endpoint in use");
+
+  ctxt.getResults().setIface(EsiDpiEndpoint<AnyPointer, AnyPointer>::Client(
+      kj::heap<EndpointServer>(*ep)));
+  return kj::READY_NOW;
+}
+
+/// ----- RpcServer definitions.
+
+RpcServer::RpcServer() : mainThread(nullptr), stopSig(false) {}
+RpcServer::~RpcServer() { stop(); }
+
+void RpcServer::mainLoop(uint16_t port) {
+  capnp::EzRpcServer rpcServer(kj::heap<CosimServer>(&endpoints),
+                               /* bindAddress */ "*", port);
+  auto &waitScope = rpcServer.getWaitScope();
+
+  // OK, this is uber hacky, but it unblocks me and isn't _too_ inefficient. The
+  // problem is that I can't figure out how read the stop signal from libkj
+  // asyncrony land.
+  //
+  // IIRC the main libkj wait loop uses `select()` (or something similar on
+  // Windows) on its FDs. As a result, any code which checks the stop variable
+  // doesn't run until there is some I/O. Probably the right way is to set up a
+  // pipe to deliver a shutdown signal.
+  //
+  // TODO: Figure out how to do this properly, if possible.
```

Comment:
i don't know anything about libkj, so unfortunately i don't have any input here.

Reply:
There's only a very small subset of devs who know anything about libkj.